pkg/storage/streams/store.go

```diff
@@ -61,43 +65,65 @@ type Store interface {
 type streamStore struct {
 	segments     segments.Store
 	segmentSize  int64
-	rootKey      *storj.Key
+	searcher     *vpath.Searcher
 	encBlockSize int
 	cipher       storj.Cipher
 }
 
 // NewStreamStore stuff
-func NewStreamStore(segments segments.Store, segmentSize int64, rootKey *storj.Key, encBlockSize int, cipher storj.Cipher) (Store, error) {
+func NewStreamStore(segments segments.Store, segmentSize int64, searcher *vpath.Searcher, encBlockSize int, cipher storj.Cipher) (Store, error) {
 	if segmentSize <= 0 {
 		return nil, errs.New("segment size must be larger than 0")
 	}
-	if rootKey == nil {
-		return nil, errs.New("encryption key must not be empty")
-	}
 	if encBlockSize <= 0 {
 		return nil, errs.New("encryption block size must be larger than 0")
 	}
 
 	return &streamStore{
 		segments:     segments,
 		segmentSize:  segmentSize,
-		rootKey:      rootKey,
+		searcher:     searcher,
 		encBlockSize: encBlockSize,
 		cipher:       cipher,
 	}, nil
 }
 
+// encryptPath finds the appropriate key for the unencrypted path and encrypts the remainder.
+func (s *streamStore) encryptPath(ctx context.Context, path storj.Path, pathCipher storj.Cipher) (encPath storj.Path, pathKey *storj.Key, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	// TODO(jeff): we can do stuff with the returned map if base is nil because it allows us
+	// to decrypt specific components. This is important during list.
+	_, consumed, base := s.searcher.Lookup(path)
+	if base == nil {
+		return "", nil, Error.New("cannot find encryption information for: %q", path)
+	}
+
+	remainingPath := path[len(consumed):]
```

Comment:
I am quite anxious about such kind of operations. Does `remainingPath` include a leading slash now?

i.e. if `path` is `a/b/c` and consumed is `a` then `remainingPath` will be `/b/c`, not `b/c`. I guess this can lead to some problems.

Reply:
Yeah, I'm also anxious about it. I'm also anxious about how sometimes a `storj.Path` has a bucket prefix and sometimes it doesn't.

I'm probably going to be doing a larger refactoring to create some type that encapsulates all of this complexity and makes it very type safe. Like a `storj.BucketPath` or something.